### PR TITLE
 stake test for create and call contract

### PR DESCRIPTION
**Pull request by cyl19970726**

### CODE REVIEW

File: tests/web3q_test.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
+package tests
```
Comment (ninjaahhh) [inline tests/web3q_test.go:1]: why do we need this file? can we simply reuse `state_test.go`?

**Review by ninjaahhh — COMMENTED** [review]
1. as mentioned above, we don't need a separate test file 
2. we only need to care about latest fork in test fixtures (London)
3. for better readability, each case with different run time code needs to have a link to its solidity source code or simply added inside the file

File: tests/web3q_test.go (hunk 1 above)
Comment (cyl19970726) [reply tests/web3q_test.go:1]: > For the one point ,I will merge web3q_test.go file into state_test.go file
> For the second point,I will delete other cases
> For the third point, I will add one more property to the test json file, key=soliditycode ,value=source code
like this:
"0x0000000000000000000000000000000000000bbb" : {
        "balance" : "0x0DE0B6B3A7640000",
        "code" : "0x608060405234801561001057600080fd5b506004361061002b5760003560e01c8063c4f038a514610030575b600080fd5b61004a600480360381019061004591906101d5565b610061565b60405161005892919061024a565b60405180910390f35b6060600080600061007185610141565b91509150806100d557600067ffffffffffffffff81111561009557610094610374565b5b6040519080825280601f01601f1916602001820160405280156100c75781602001600182028036833780820191505090505b50600093509350505061013c565b60008267ffffffffffffffff8111156100f1576100f0610374565b5b6040519080825280601f01601f1916602001820160405280156101235781602001600182028036833780820191505090505b5090506000838160208401893c81600195509550505050505b915091565b600080600073ffffffffffffffffffffffffffffffffffffffff168373ffffffffffffffffffffffffffffffffffffffff16141561018557600080915091506101bb565b60008060009050843b9150808210156101a6576000809350935050506101bb565b80826101b29190610296565b60019350935050505b915091565b6000813590506101cf816103b9565b92915050565b6000602082840312156101eb576101ea6103a3565b5b60006101f9848285016101c0565b91505092915050565b61020b816102dc565b82525050565b600061021c8261027a565b6102268185610285565b9350610236818560208601610312565b61023f816103a8565b840191505092915050565b600060408201905081810360008301526102648185610211565b90506102736020830184610202565b9392505050565b600081519050919050565b600082825260208201905092915050565b60006102a182610308565b91506102ac83610308565b9250828210156102bf576102be610345565b5b828203905092915050565b60006102d5826102e8565b9050919050565b60008115159050919050565b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b6000819050919050565b60005b83811015610330578082015181840152602081019050610315565b8381111561033f576000848401525b50505050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fd5b600080fd5b6000601f19601f8301169050919050565b6103c2816102ca565b81146103cd57600080fd5b5056fea26469706673582212208771ce4b96274bcaae49feedb8cea361ea779c0e1ea213011b5d61735675975364736f6c63430008070033",
        "nonce" : "0x00",
        "storage" : {
        },
        "soliditycode" : "contract TestCall{\n\n    function sizeRaw(address addr) internal view returns (uint256, bool) {\n        if (addr == address(0x0)) {\n            return (0, false);\n        }\n        uint256 codeSize;\n        uint256 off = 0;\n        assembly {\n            codeSize := extcodesize(addr)\n        }\n        if (codeSize < off) {\n            return (0, false);\n        }\n\n        return (codeSize - off, true);\n    }\n\n    function getRaw(address addr) public view returns (bytes memory, bool) {\n        (uint256 dataSize, bool found) = sizeRaw(addr);\n\n        if (!found) {\n            return (new bytes(0), false);\n        }\n\n        // copy the data without the \"code\"s\n        bytes memory data = new bytes(dataSize);\n        uint256 off = 0;\n        assembly {\n            // retrieve data size\n            extcodecopy(addr, add(data, 0x20), off, dataSize)\n        }\n        return (data, true);\n    }\n\n}"
      }

File: tests/state_test.go [hunk 2]
```diff
@@ -131,3 +135,60 @@ func withTrace(t *testing.T, gasLimit uint64, test func(vm.Config) error) {
...
+		})
+	}
+}
+
+func StateTrie(db *state.StateDB, test *StateTest, t *testing.T) {
+	noContractCreation := test.json.Tx.To != ""
+
+	fmt.Println("--------------------state info-------------------")
+	for addr, _ := range test.json.Pre {
+		//object := db.GetOrNewStateObject(addr)
```
Comment (ninjaahhh) [inline tests/state_test.go:173]: remove unused comment

File: tests/state_test.go [hunk 3]
```diff
@@ -131,3 +135,60 @@ func withTrace(t *testing.T, gasLimit uint64, test func(vm.Config) error) {
...
+				})
+			}
+		})
+	}
+}
+
+func StateTrie(db *state.StateDB, test *StateTest, t *testing.T) {
+	noContractCreation := test.json.Tx.To != ""
+
+	fmt.Println("--------------------state info-------------------")
```
Comment (ninjaahhh) [inline tests/state_test.go:171]: use `t.Log`

File: tests/state_test.go [hunk 4]
```diff
@@ -131,3 +135,60 @@ func withTrace(t *testing.T, gasLimit uint64, test func(vm.Config) error) {
...
+						t.Error(err)
+					}
+
+				})
+			}
+		})
+	}
+}
+
+func StateTrie(db *state.StateDB, test *StateTest, t *testing.T) {
```
Comment (ninjaahhh) [inline tests/state_test.go:168]: `StateTrie` => `printStateTrie`, we don't want to export this method

File: tests/state_test_util.go [hunk 5]
```diff
@@ -230,8 +232,18 @@ func (t *StateTest) RunNoVerify(subtest StateSubtest, vmconfig vm.Config, snapsh
 	snapshot := statedb.Snapshot()
 	gaspool := new(core.GasPool)
 	gaspool.AddGas(block.GasLimit())
-	if _, err := core.ApplyMessage(evm, msg, gaspool); err != nil {
```
Comment (ninjaahhh) [inline tests/state_test_util.go:233]: as discussed offline, we can directly print the error here

Comment (cyl19970726) [reply tests/state_test_util.go:233]: has been fixed

**Review by ninjaahhh — COMMENTED** [review]
looks good to me. remember to change the base branch to `tm_w3q`

**Pull request closed without merging**